### PR TITLE
k8sclusterreceiver: Ensure informer caches are synced before initial data sync

### PR DESCRIPTION
**Pull request by asuresh4**
**Description:** Wait for informer caches to sync before the initial data sync happens. Otherwise, it is possible that the resource handler callbacks process events before the caches have synced which leads to missing data in some cases when relying on the informer store for data.

**Testing:** Updated tests.

### CODE REVIEW

File: receiver/k8sclusterreceiver/watcher.go
Comment (dmitryax) [inline receiver/k8sclusterreceiver/watcher.go:46]: Why it has to be a boolean flag? Can't we synchronize it with a channel which would block `kr.resourceWatcher.startWatchingResources` until the cache is synced?

Comment (asuresh4) [reply receiver/k8sclusterreceiver/watcher.go:46]: I tried that approach initially but I learnt that `informers.SharedInformerFactory.Start` needs to be invoked first for the cache sync to begin. And as soon as `Start` is invoked events could come through to the Resource Event Handler before the caches are synced. So in this particular case the receiver would have to invoke `rw.sharedInformerFactory.Start(stopper)` to start kick-off the sync and ensure any events that come through before the initial sync is complete, blocks until the initial cache sync is complete or it times out.

File: receiver/k8sclusterreceiver/collection/pods.go
Comment (dmitryax) [inline receiver/k8sclusterreceiver/collection/pods.go:217]: let's avoid going too far in DRY :)
looking at `handleErrors(err, rsRef, pod.UID, logger)` I would never expect that this function write warnings under assumption it's called with `exist=false`. 
It'd be much more readable if all the error/warning stuff stays in this function

Comment (asuresh4) [reply receiver/k8sclusterreceiver/collection/pods.go]: Updated to make this less cryptic. Let me know what you think.

File: receiver/k8sclusterreceiver/watcher.go
Comment (dmitryax) [inline receiver/k8sclusterreceiver/watcher.go:166]: Why do we need this timeout? What if we just keep waiting for the cache as long as needed? We will not start updating dimensions until that, right? Isn't it desired behavior? Why would we need to send "corrupted" updates if we wasn't able to wait for the cache?

Comment (asuresh4) [reply receiver/k8sclusterreceiver/watcher.go:166]: I was thinking about this as well since we always want to make sure the data is accurate before starting to sync. I've removed this timeout.

**Review by dmitryax — APPROVED** [review]
LGTM

File: receiver/k8sclusterreceiver/watcher.go
Comment (tigrannajaryan) [inline receiver/k8sclusterreceiver/watcher.go:161]: This can keep running infinitely. Should we have a timeout to avoid getting stuck forever if something goes wrong?

Comment (asuresh4) [reply receiver/k8sclusterreceiver/watcher.go:161]: I originally had added a timeout here. But then later removed it since we want to have ensure the we have all the data before the sync begins. Also, playing around with this, so far, it looks like the caches are synced in < 20s for our lab deployment (however, might vary based on size of the deployment). We could have a timeout here to safeguard against the risk of running infinitely, but that would potentially mean we sync incomplete metadata during the initial sync. On subsequent metadata updates to the relevant dimension (or label in OTel), if the cache is synced, the missed out metadata will also be synced but until then the synced metadata would be incomplete. @tigrannajaryan @dmitryax

Comment (tigrannajaryan) [reply receiver/k8sclusterreceiver/watcher.go:161]: I don't think we want to have infinite loops that are not cancellable in the Collector. This will create problems for graceful shutdown in the future too. We should probably add cancellation support at least.

Comment (asuresh4) [reply receiver/k8sclusterreceiver/watcher.go:161]: I agree, will make an update to handle this better.

Comment (dmitryax) [reply receiver/k8sclusterreceiver/watcher.go:161]: @asuresh4 correct me if I'm wrong here, but as I understood the original timeout was implemented in a "best effort" way. If cache sync is not complete we ignore parts of the metadata that were not able to sync. That actually causes corrupted metadata updates in the current version of collector. We should avoid it, but I'm not sure what's the best way. Probably better to fail to start the receiver. But if we don't do anything it'll be handled by liveness probes anyway.

Comment (dmitryax) [reply receiver/k8sclusterreceiver/watcher.go:161]: I'd suggest the following: 
1. make the cache sync not blocking the pipeline building (not sure if it is already);
2. add cancellation support;
3. have a reasonable default timeout for cache sync, like 2-5 minutes;
2. by timeout, print an error message and deactivate the receiver.
Not sure how feasible is this. @tigrannajaryan @asuresh4 what do you think?

Comment (tigrannajaryan) [reply receiver/k8sclusterreceiver/watcher.go:161]: Discussed offline, accepting @dmitryax's suggestion, except in step 4 we call ReportFatalError and terminate the Collector since we believe this is critical because there is only one instance of Collector that is responsible for this collection.

Comment (asuresh4) [reply receiver/k8sclusterreceiver/watcher.go]: Updated as per our discussion. I've set the timeout to be 10 minutes (which I believe should be long enough for all cases), after which the receiver reports a fatal error. Also, leveraged cancellation support of `rw.sharedInformerFactory.WaitForCacheSync` to make sure it returns if the receiver is Shutdown while it's still waiting for the sync. PTAL @dmitryax @tigrannajaryan.

File: receiver/k8sclusterreceiver/watcher.go
Comment (dmitryax) [inline receiver/k8sclusterreceiver/watcher.go:112]: would be `defer cancel()` better instead?

**Review by dmitryax — APPROVED** [review]
LGTM

File: receiver/k8sclusterreceiver/watcher.go
Comment (asuresh4) [reply receiver/k8sclusterreceiver/watcher.go:112]: In this case since it's the last line to be executed within the scope of the method, I believe it shouldn't make a difference.

Comment (dmitryax) [reply receiver/k8sclusterreceiver/watcher.go:112]: no difference, just for consistency with general practice. anyway your call

Comment (asuresh4) [reply receiver/k8sclusterreceiver/watcher.go]: 👍 updated.

**Comment by asuresh4** [comment]
@tigrannajaryan - please let me know if you have further comments. Would like to get this merged soon.

**Comment by tigrannajaryan** [comment]
@asuresh4 please resolve the conflicts.

**Review by tigrannajaryan — APPROVED** [review]
Good to merge once CircleCI build passes.

**Pull request merged**
